rust/lance/src/dataset/optimize.rs

```diff
@@ -1585,4 +1615,55 @@ mod tests {
         // and so we just see the bump from the commit_compaction
         assert_eq!(dataset.manifest.version, 6);
     }
+
+    #[test]
+    fn test_missing_ids() {
+        // test with missing first row
+        // test with missing last row
+        // test fragment ids out of order
+
+        let fragments = vec![
+            Fragment {
+                id: 0,
+                files: Vec::new(),
+                deletion_file: None,
+                physical_rows: 5,
```

Comment:
All I have to do to make this test fail is set `physical_rows` to zero. So Rob's assessment was correct that this alone caused the hang.

Reply:
nit: let's add a test case with `physical_rows==0` for `remap_file`?

Reply:
There's not an easy way to do that, since we can't write a dataset that writes fragments the old way with our current API.

Reply:
I see. Nvm on this. I guess we should start adding some backward compatibility test utils.